src/Sarif/Readers/DeferredDictionary.cs

```diff
@@ -0,0 +1,514 @@
+            }
+
+            public KeyValuePair<string, U> Current { get; private set; }
+            object IEnumerator.Current => Current;
+
+            public void Dispose()
+            {
+                if (_stream != null)
+                {
+                    _stream.Dispose();
+                    _stream = null;
+                    _jsonTextReader = null;
+                }
+            }
+
+            public bool MoveNext()
+            {
+                if (_jsonTextReader.TokenType == JsonToken.EndObject) return false;
+
+                // Read the next key
+                string key = (string)_jsonTextReader.Value;
+
+                // Read the value
+                _jsonTextReader.Read();
+                U value = _jsonSerializer.Deserialize<U>(_jsonTextReader);
+
+                // Read EndObject, next is StartObject of next member
+                _jsonTextReader.Read();
+
+                Current = new KeyValuePair<string, U>(key, value);
+                return true;
+            }
+
+            public void Reset()
+            {
+                Dispose();
+
+                // Open a new Stream
+                _stream = _streamProvider();
+
+                // Seek to the object start
+                _stream.Seek(_start, SeekOrigin.Begin);
+
+                // Build a JsonTextReader
+                _jsonTextReader = new JsonTextReader(new StreamReader(_stream));
+
+                // StartObject
+                _jsonTextReader.Read();
+
+                // PropertyName of first item key
+                _jsonTextReader.Read();
+            }
+        }
+
+        /// <summary>
+        ///  ReadOnlyCollectionAdapter wraps a IEnumerator and Count getters
+        ///  to implement ICollection.
+        /// </summary>
+        /// <typeparam name="U">Collection Item type</typeparam>
+        private class ReadOnlyCollectionAdapter<U> : ICollection<U>
+        {
+            private Func<IEnumerator<U>> _enumeratorFactory;
+            private Func<int> _countGetter;
+            public int Count => _countGetter();
+            public bool IsReadOnly => true;
+
+            public ReadOnlyCollectionAdapter(Func<IEnumerator<U>> enumeratorFactory, Func<int> countGetter)
+            {
+                _enumeratorFactory = enumeratorFactory;
+                _countGetter = countGetter;
+            }
+
+            #region Mutators [Not Supported]
+            public void Add(U item)
+            {
+                throw new NotSupportedException();
+            }
+
+            public void Clear()
+            {
+                throw new NotSupportedException();
+            }
+
+            public bool Remove(U item)
+            {
+                throw new NotSupportedException();
+            }
+            #endregion
+
+            public bool Contains(U item)
+            {
+                EqualityComparer<U> comparer = EqualityComparer<U>.Default;
+
+                foreach(U value in this)
```

Comment:
>foreach [](start = 16, length = 7)

here and elsewhere need space before parens, see foreach and if statements in this class

Reply:
Did a format document.